### DISCUSSION TITLE
ServiceStack UI and Complex Types

### DISCUSSION THREAD

**Discussion by writeameer** [ServiceStack/Discuss#18] [Q&A] [answered]
Hi there Everyone,

Noob here, beginning to learn ServiceStack. I have a service with the following request response objects:

```
[Route("/Azure/CreateDc")]
public class CreateAzureDcRequest: IReturn<CreateAzureDcResponse>
{
    public string Location { get; set; }
}

public class CreateAzureDcResponse
{
    public string KGuid { get; set; }
    public string Status { get; set; }
}

```

The above service functions as expected and I can see the ServiceStack UI (The Swagger like browser interface) responds as appropriate. Now when I make the **request type** slightly more complex, the ServiceStack UI freezes. Here's an example:

```
[Route("/Azure/CreateDc")]
public class CreateAzureDcRequest: IReturn<CreateAzureDcResponse>
{
    public ComplexType complexType { get; set; }
}

public class CreateAzureDcResponse
{
    public string KGuid { get; set; }
    public string Status { get; set; }
}

public class ComplexType
{
    public string Location { get; set; }
}
```

The UI just freezes and I'm unable to interact with the API:

![image](https://user-images.githubusercontent.com/221149/161542295-414ea967-caa5-4969-a3c0-e2bcfb369153.png)

Do the DTO members have to be a primitive/native type ? Can we not have more complex types when making API calls ? Thanks for any pointers in the right directions !

Sincerely,
Ameer Deen.

**Comment by mythz** [comment, marked as answer]
It's not freezing, it just can't display a form UI control that can submit a complex type property so it's ignored, i.e. there's no HTML input control that can send a complex type so it's ignored in the UI. You'll still be able to call the API through code, you just wont be able to get an Auto UI that will be able to populate complex type properties.

You'll need to use a flat Request DTO structure to be included in an Auto Form UI.

**Reply by writeameer** [reply]
Thanks @mythz  - You were spot on - much appreciated !
